File: src/main/java/com/distribution/modules/card/service/impl/CardInfoServiceImpl.java
package com.distribution.modules.card.service.impl;

import com.alibaba.fastjson.JSON;
import com.distribution.common.utils.OkHttpUtil;
import com.distribution.modules.api.dao.UserDao;
import com.distribution.modules.card.dao.CardInfoMapper;
import com.distribution.modules.card.entity.CardApiResponse;
import com.distribution.modules.card.entity.CardInfo;
import com.distribution.modules.card.service.CardInfoService;
import com.distribution.modules.dis.entity.DisMemberInfoEntity;
import okhttp3.Request;
import okhttp3.Response;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author ChunLiang Hu
 * @Company
 * @Project distribution
 * @Package com.distribution.modules.card.service.impl
 * @Description TODO(描述)
 * @create 2018/5/8-20:23
 */
@Service
public class CardInfoServiceImpl implements CardInfoService {
    @Autowired
    private CardInfoMapper cardInfoMapper;
    @Autowired
    private UserDao userDao;

    /**
     * 根据ID查询
     *
     * @param id
     * @return
     */
    @Override
    public CardInfo queryObject(String id) {
        return cardInfoMapper.selectByPrimaryKey(id);
    }

    /**
     * 查询列表
     *
     * @param map
     * @return
     */
    @Override
    public List<CardInfo> queryList(Map<String, Object> map) {
        return cardInfoMapper.selectList(map);
    }

    /**
     * 保存
     *
     * @param cardInfo
     * @throws Exception
     */
    @Override
    public void save(CardInfo cardInfo) throws Exception {
        if (StringUtils.isBlank(cardInfo.getBankNum())) {
            return;
        }
        CardApiResponse apiResponse = getProductInfo(cardInfo.getBankNum());
        if (apiResponse.isSuccess()) {
            cardInfo.setCardImg(apiResponse.getResults().getBackgroundImg());
            cardInfo.setCardUrl(apiResponse.getResults().getLink());
            cardInfoMapper.insertSelective(cardInfo);
        }
    }

    /**
     * 更新
     *
     * @param cardInfo
     * @throws Exception
     */
    @Override
    public void update(CardInfo cardInfo) throws Exception {
        if (StringUtils.isBlank(cardInfo.getBankNum())) {
            return;
        }
        CardApiResponse apiResponse = getProductInfo(cardInfo.getBankNum());
        if (apiResponse.isSuccess()) {
            cardInfo.setCardImg(apiResponse.getResults().getIcon());
            cardInfo.setCardUrl(apiResponse.getResults().getLink());
            cardInfoMapper.updateByPrimaryKeySelective(cardInfo);
        }
    }

    /**
     * 删除
     *
     * @param id
     * @throws Exception
     */
    @Override
    public void delete(String id) throws Exception {
        cardInfoMapper.deleteByPrimaryKey(id);
    }

    /**
     * 批量删除
     *
     * @param ids
     * @throws Exception
     */
    @Override
    public void deleteBatch(String[] ids) throws Exception {
        cardInfoMapper.deleteBatch(ids);
    }

    @Override
    public CardApiResponse getProductInfo(String prodId) throws Exception {

        String url = OkHttpUtil.attachHttpGetParam("http://www.qichangkeji.vip/qckjgzhManager/DownSingleLoan/selectById.do",
                "id", prodId);
        Request request = new Request.Builder().url(url).get().build();
        Response response = OkHttpUtil.execute(request);
        if (response.isSuccessful()) {
            return JSON.parseObject(response.body().string(), CardApiResponse.class);
        }
        return null;
    }

    @Override
    public String getProductUrl(DisMemberInfoEntity member, String prodId) throws Exception {
        CardInfo cardInfo = cardInfoMapper.selectByBankNum(prodId);
        Map<String, String> params = new HashMap<>();
        params.put("name", member.getDisUserName());
        params.put("phone", member.getUserEntity().getMobile());
        params.put("type", "1");
        params.put("goodsId", prodId);
        params.put("idCard", member.getIdCode());
        params.put("fatherId", "5710");
        params.put("otherUserId", "0");
        String parentMobile = "0";
        if (member.getDisMemberParent() != null) {
            parentMobile = userDao.queryByMemberId(member.getDisMemberParent().getId()).getMobile();
            params.put("otherUserId", parentMobile);
        }
        String url = OkHttpUtil.attachHttpGetParams("http://www.qichangkeji.vip/qckjgzhManager/DownUser/Add.do",
                params);
//        RequestBody formBody = RequestBody.create(MediaType.parse("text/json; charset=utf-8"), JSON.toJSONString(params));
//        Request request = new Request.Builder().url("http://www.qichangkeji.vip/qckjgzhManager/DownUser/Add.do")
//                .post(formBody).build();
        Request request = new Request.Builder().url(url).get().build();
        try (Response response = OkHttpUtil.execute(request)) {
            System.out.println(response.body().string());
            return response.isSuccessful() ? cardInfo.getCardUrl() : null;
        }
    }
}

File: src/main/kotlin/com/distribution/modules/card/service/impl/CardInfoServiceImpl.kt
package com.distribution.modules.card.service.impl

import com.alibaba.fastjson.JSON
import com.distribution.common.utils.OkHttpUtil
import com.distribution.modules.card.dao.CardInfoMapper
import com.distribution.modules.card.entity.CardApiResponse
import com.distribution.modules.card.entity.CardInfo
import com.distribution.modules.card.service.CardInfoService
import com.distribution.modules.dis.entity.DisMemberInfoEntity
import com.distribution.modules.dis.service.DisFansService
import okhttp3.Request
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*

/**
 * @author ChunLiang Hu
 * @Company
 * @Project distribution
 * @Package com.distribution.modules.card.service.impl
 * @Description TODO(描述)
 * @create 2018/5/8-20:23
 */

@Service
class CardInfoServiceImpl : CardInfoService {
    @Autowired
    private lateinit var cardInfoMapper: CardInfoMapper
    @Autowired
    private lateinit var fansService: DisFansService
    private val log = LoggerFactory.getLogger(CardInfoServiceImpl::class.java)

    /**
     * 根据ID查询
     *
     * @param id
     * @return
     */
    override fun queryObject(id: String): CardInfo {
        return cardInfoMapper.selectByPrimaryKey(id)
    }

    override fun queryByBankNum(bankNum: String): CardInfo {
        return cardInfoMapper.selectByBankNum(bankNum)
    }

    /**
     * 查询列表
     *
     * @param map
     * @return
     */
    override fun queryList(map: Map<String, Any>): List<CardInfo> {
        return cardInfoMapper.selectList(map)
    }

    /**
     * 保存
     *
     * @param cardInfo
     * @throws Exception
     */
    @Throws(Exception::class)
    override fun save(cardInfo: CardInfo) {
        if (StringUtils.isBlank(cardInfo.bankNum)) {
            return
        }
        when {
            cardInfo.cardImg.isNotEmpty() && cardInfo.cardUrl.isNotEmpty() -> cardInfoMapper.insertSelective(cardInfo)
            else -> {
                val apiResponse = getProductInfo(cardInfo.bankNum)
                if (apiResponse.isSuccess) {
                    cardInfo.cardImg = apiResponse.results.backgroundImg
                    cardInfo.cardUrl = apiResponse.results.link
                    cardInfoMapper.insertSelective(cardInfo)
                }
            }
        }
    }

    /**
     * 更新
     *
     * @param cardInfo
     * @throws Exception
     */
    @Throws(Exception::class)
    override fun update(cardInfo: CardInfo) {
        if (StringUtils.isBlank(cardInfo.bankNum)) {
            return
        }
        val apiResponse = getProductInfo(cardInfo.bankNum)
        if (apiResponse.status == "1") {
            cardInfo.cardImg = apiResponse.results.icon
            if (cardInfo.useThirdPart == 1) {
                cardInfo.cardUrl = apiResponse.results.link
            }
            cardInfoMapper.updateByPrimaryKeySelective(cardInfo)
        }
    }

    /**
     * 删除
     *
     * @param id
     * @throws Exception
     */
    @Throws(Exception::class)
    override fun delete(id: String) {
        cardInfoMapper.deleteByPrimaryKey(id)
    }

    /**
     * 批量删除
     *
     * @param ids
     * @throws Exception
     */
    @Throws(Exception::class)
    override fun deleteBatch(ids: Array<String>) {
        cardInfoMapper.deleteBatch(ids)
    }

    @Throws(Exception::class)
    override fun getProductInfo(prodId: String): CardApiResponse {

        val url = OkHttpUtil.attachHttpGetParams("http://www.qichangkeji.vip/qckjgzhManager/DownSingleLoan/selectById.do",
                mapOf("id" to prodId, "fatherId" to "5710"))
        val request = Request.Builder().url(url).get().build()
        val response = OkHttpUtil.execute(request)
        return if (response.isSuccessful) {
            JSON.parseObject(response.body()!!.string(), CardApiResponse::class.java)
        } else CardApiResponse()
    }

    @Throws(Exception::class)
    override fun getProductUrl(member: DisMemberInfoEntity, prodId: String): String {
        val cardInfo = cardInfoMapper.selectByBankNum(prodId)
        //如果不推送数据直接返回URL
        if (cardInfo.useThirdPart == 0) {
            return cardInfo.cardUrl
        }
        val params = HashMap<String, String>()
        params["name"] = member.disUserName!!
        params["phone"] = member.userEntity!!.mobile!!
        params["type"] = "1"
        params["goodsId"] = prodId
        params["idCard"] = member.idCode!!
        params["fatherId"] = "5710"
        val fans = fansService.queryByOpenId(member.openId!!)!!
        params["otherUserId"] = fans.workerId.toString()
        val url = OkHttpUtil.attachHttpGetParams("http://www.qichangkeji.vip/qckjgzhManager/DownUser/Add.do", params)
        //        RequestBody formBody = RequestBody.create(MediaType.parse("text/json; charset=utf-8"), JSON.toJSONString(params));
        //        Request request = new Request.Builder().url("http://www.qichangkeji.vip/qckjgzhManager/DownUser/Add.do")
        //                .post(formBody).build();
        val request = Request.Builder().url(url).get().build()
        OkHttpUtil.execute(request).use { response ->
            log.info(response.body()!!.string())
            return if (response.isSuccessful) cardInfo.cardUrl else ""
        }
    }
}
